clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3582,21 +3407,55 @@ static bool InvalidWorkGroupSizeAttrs(const Expr *MGValue, const Expr *XDim,
            ZDimExpr->getResultAsAPSInt() != 1));
 }
 
-// If the [[intel::max_work_group_size(X, Y, Z)]] attribute is specified on
-// a declaration along with [[sycl::reqd_work_group_size(X1, Y1, Z1)]]
-// attribute, check to see if values of reqd_work_group_size arguments are
+// If the 'reqd_work_group_size' attribute is specified on
+// a declaration along with 'max_work_group_size' attribute,
+// check to see if values of reqd_work_group_size arguments are
 // equal or less than values of max_work_group_size attribute arguments.
-static bool checkWorkGroupSizeAttrValues(const Expr *RWGS, const Expr *MWGS) {
+// If the 'reqd_work_group_size' attribute is specified on
+// a declaration along with 'reqd_work_group_size' attribute,
+// check to see if values of reqd_work_group_size arguments are
+// equal or greater than values of 'reqd_work_group_size' attribute arguments.
+
+template <typename Comparator>
+static bool checkWorkGroupSizeAttrValues(
+    Sema &S, const Expr *RWGSXDim, const Expr *RWGSYDim, const Expr *RWGSZDim,
+    const Expr *WGSXDim, const Expr *WGSYDim, const Expr *WGSZDim) {
   // If any of the operand is still value dependent, we can't test anything.
-  const auto *RWGSCE = dyn_cast<ConstantExpr>(RWGS);
-  const auto *MWGSCE = dyn_cast<ConstantExpr>(MWGS);
-
-  if (!RWGSCE || !MWGSCE)
+  const auto *RWGSXDimExpr = dyn_cast<ConstantExpr>(RWGSXDim);
+  const auto *RWGSYDimExpr = dyn_cast<ConstantExpr>(RWGSYDim);
+  const auto *RWGSZDimExpr = dyn_cast<ConstantExpr>(RWGSZDim);
+  const auto *WGSXDimExpr = dyn_cast<ConstantExpr>(WGSXDim);
+  const auto *WGSYDimExpr = dyn_cast<ConstantExpr>(WGSYDim);
+  const auto *WGSZDimExpr = dyn_cast<ConstantExpr>(WGSZDim);
+
+  if (!RWGSXDimExpr || !RWGSYDimExpr || !RWGSZDimExpr || !WGSXDimExpr ||
+      !WGSYDimExpr || !WGSZDimExpr)
     return false;
 
-  // Otherwise, check if value of reqd_work_group_size argument is
-  // greater than value of max_work_group_size attribute argument.
-  return RWGSCE->getResultAsAPSInt() > MWGSCE->getResultAsAPSInt();
+  // Otherwise, check if value of 'reqd_work_group_size' attribute argument
+  // is greater than value of 'max_work_group_size' attribute argument.
+  // or check if value of 'reqd_work_group_size' attribute argument is less
+  // than value of 'reqd_work_group_size' attribute argument.
```

Comment:
```
// The second set of values may be either 'max_work_group_size' or 'reqd_work_group_size' attribute arguments.
// Compare (using the comparison operator passed in) appropriately with the first set of 'reqd_work_group_size' arguments.
```

To me, describing it this way seems to make sense.  I am not sure if @AaronBallman or other reviewers share the same opinion.  :-)  Please wait to hear from others before making this change.   I am not too tied to any of my suggestions specifically, but I do find the current comments (particularly about multiple 'reqd_work_group_size' values confusing).

Reply:
I did not make any changes to the comments. I will wait for response from @AaronBallman or other reviewers before making this change.

Reply:
I have updated comments.

@premanandrao, @AaronBallman could you please take a look at the patch again. Thanks.